resource_elasticsearch_xpack_role_mapping.go

```diff
@@ -0,0 +1,232 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	elastic6 "gopkg.in/coveo/elasticsearch-client-go.v6"
+	elastic5 "gopkg.in/olivere/elastic.v5"
+)
+
+func resourceElasticsearchXpackRoleMapping() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceElasticsearchXpackRoleMappingCreate,
+		Read:   resourceElasticsearchXpackRoleMappingRead,
+		Update: resourceElasticsearchXpackRoleMappingUpdate,
+		Delete: resourceElasticsearchXpackRoleMappingDelete,
+
+		Schema: map[string]*schema.Schema{
+			"role_mapping_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"enabled": &schema.Schema{
+				Type:     schema.TypeBool,
+				Default:  true,
+				Optional: true,
+			},
+			"rules": &schema.Schema{
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: suppressEquivalentJson,
+			},
+			"roles": &schema.Schema{
+				Type: schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Required: true,
+			},
+			"metadata": &schema.Schema{
+				Type:             schema.TypeString,
+				Optional:         true,
+				Default:          "{}",
+				DiffSuppressFunc: suppressEquivalentJson,
+			},
+		},
+	}
+}
+
+func resourceElasticsearchXpackRoleMappingCreate(d *schema.ResourceData, m interface{}) error {
+	name := d.Get("role_mapping_name").(string)
+
+	reqBody, err := buildPutRoleMappingBody(d, m)
+	err = xpackPutRoleMapping(d, m, name, reqBody)
+	if err != nil {
+		return err
+	}
+	d.SetId(name)
+	return resourceElasticsearchXpackRoleMappingRead(d, m)
+}
+
+func resourceElasticsearchXpackRoleMappingRead(d *schema.ResourceData, m interface{}) error {
+
+	roleMapping, err := xpackGetRoleMapping(d, m, d.Id())
+	if err != nil {
+		fmt.Println("Error during read")
+		if elasticErr, ok := err.(*elastic6.Error); ok && elasticErr.Status == 404 {
+			fmt.Printf("[WARN] Role mapping %s not found. Removing from state\n", d.Id())
+			d.SetId("")
+			return nil
+		}
+		if elasticErr, ok := err.(*elastic5.Error); ok && elasticErr.Status == 404 {
```

Comment:
Instead of checking everywhere with the elastic5 client. You could add a function called in `Create` and `Read` (covers all entry points)
```
func validateElasticSix(m interface{}) error {
	if client, ok := m.(*elastic5.Client); ok {
		return errors.New("unsupported in elasticv5 client")
	}
        return nil
}
```
After, you can assume everywhere that the client is elastic6

Reply:
Wasnt too sure about this one, I didn't change it as I prefer the extensibility this provides. If someone wants to add the v5 support he won't have to change the code, just implement the function.